server/service/handler.go

```diff
@@ -1032,6 +1035,47 @@ func RedirectSetupToLogin(svc fleet.Service, logger kitlog.Logger, next http.Han
 	}
 }
 
+func WithDEPWebviewRedirect(svc fleet.Service, logger kitlog.Logger, next http.Handler, urlPrefix string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/mdm/sso" && r.URL.RawQuery == "" {
+			level.Debug(logger).Log("msg", "handling mdm sso", "url", r.URL.String())
+			// Note: We'll apply this redirect only if query params are empty because want to
+			// redirect to the same URL with added query params after parsing the x-apple-aspen-deviceinfo
+			// header. Whenever we see a request with any query params already present, we'll
+			// skip this step and just continue to the next handler.
+
+			// TODO: should we do any precautionary checks befere we try to parse the header?
+			// xml or pki performance concerns?
+			di := r.Header.Get("X-apple-aspen-deviceinfo")
+			if di != "" {
+				level.Debug(logger).Log("msg", "parsing X-apple-aspen-deviceinfo", "url", r.URL.String())
+				// extract x-apple-aspen-deviceinfo custom header from request
+				_, err := dep_webview.DefaultParser.Parse(r)
```

Comment:
Same here, if we extract just the parser logic as a function, we can then pass it the `di` variable (the device info header value) directly.